### ISSUE TITLE
NicSetSetting: change of MAC Address does not take effect until manual restart of VPN Client

### ISSUE THREAD

**Issue by scandox** [SoftEtherVPN/SoftEtherVPN#172]
I have been using NicSetSetting to set the MAC address of Virtual NICs joining a Hub. I used static leases, so the MAC address effectively determines the IP Address the NIC should obtain. 

I have noticed however that the change of MAC Address, though recorded in Softether data, is not reflected in the actual OS, until after a restart of the VPN Client itself. This means on first contact my NIC's are configured with incorrect IP addresses.

I also tried:

*   Enable and Disable of NIC
*   Setting it multiple times
*   Disconnecting any Accounts and then Reconnecting.

However only `vpnclient restart` works for me. I am using two separate clients:

*   Version 4.17 Build 9562 on Debian Wheezy
*    Version 4.18 Build 9570 on Raspbian

Both exhibit the same behaviour. Is this a bug or a feature?

**Comment by moatazelmasry2** [comment]
Could you please verify that this problem still occur?

**Comment by scandox** [comment]
@moatazelmasry2 Sorry for the slow response:

- I setup a server and client
- Connected the client
- Checked the MAC Address which was 00:AC:B2:5A:26:93
- Logged in to vpncmd and ran "NicSetSetting ether MAC:/00-AC-BA-B5-B4-78
- Logged out of vpncmd and ran ifconfig. Still seeting MAC ending with 93
- Restarted vpnclient
- Ran ifconfig and now see MAC ending with 78

So it seems to me this issue is still live. The versions of Server and Client were:

- Client Ver 4.24, Build 9651 on Ubuntu 16.06
- Server Ver 4.24, Build 9651 on Ubuntu 16.06

**Comment by scandox** [comment]
@moatazelmasry2 I'm not really a C++ person but it seems to me the issue lies in or around the code here:

https://github.com/SoftEtherVPN/SoftEtherVPN/blob/bed99f9a56e29a0fcd7a9e3d02f84f9d8621eb3d/src/Cedar/Client.c#L8104

`r` in this case being the result of `r = Search(c->UnixVLanList, &t);`. After this there are calls to `CiSaveConfigurationFile` and `CiNotify` and `CiSendGlobalPulse` - none of which seem to do anything much at OS level. So maybe it is merely setting the MAC on the internal list, saving it to a config and then not actually attempting to update the virtual adaptor?

**Comment by moatazelmasry2** [comment]
Didn't test that yet, but this looks like a very good analysis. Solid work!!!. Now we need someone to implement that :)

**Comment by scandox** [comment]
@moatazelmasry2 I made a small POC for this and tested it on Linux. It works.

https://github.com/scandox/SoftEtherVPN/commit/73cb7a3bd5c84ddc3acb2357fc5ebc4599708f25

I think though that the design is not perhaps correct. The `UnixCreateTapDevice` function is not called from any external module, so calling `UnixModifyTapDevice` externally feels wrong. I'm also far from sure of the precise relationship between a VLAN and a tap device, within the Softether terminology.

In addition there still requires work for MAC_OS and Solaris in order to be consistent - at least for Unixlikes. Still this code does address the problem I was facing.

**Comment by Hamilkar247** [comment]
If i good understand, problem with set MAC address still exist?

**Comment by scandox** [comment]
@Hamilkar247 Yes it looks that way. There is no apparent change to the way this works,
